### PR TITLE
new way of making images

### PR DESCRIPTION
**Pull request by mollux**
This PR provided a new way of generating Docker images for Mautic 5.

The main differences with the previous M4 version are:

* clearer separation of concern: e.g. the image is no longer responsible to create the database, crons can be configured independenty from the 
* immutable images by default: no more copying over folders when starting a container based on the image
* standard support for the new messenger concept: standard doctrine queueus will be used and all queues will be consumed by workers
* separate containers for web, cron and workers so the different roles can be split up and scaled separately
* amd64 and arm64 builds

See the readme for more context, and the the `examples` folder contains ready-to-use.

### CODE REVIEW

**Comment by LyVanBong** [comment]
why not update image

File: common/entrypoint_mautic_web.sh
Comment (ilkkao) [inline common/entrypoint_mautic_web.sh:5]: could have here `echo "Waiting Mysql to be ready..." inside the loop. That would help in quite common error case where BD credentials are wrong or missing.

File: examples/basic/docker-compose.yml
Comment (ilkkao) [inline examples/basic/docker-compose.yml:59]: Doesn't cron and worker container also need MYSQL env variables? Or is that covered somehow?

Comment (mollux) [reply examples/basic/docker-compose.yml:59]: @ilkkao they are not needed, as they (should) purely execute console commands from Mautic itself, which has all the needed context.

Those are examples, so if you need thos for a non-standard script, you can always add those on your own docker-compose or similar setup

File: common/entrypoint_mautic_web.sh
Comment (mollux) [reply common/entrypoint_mautic_web.sh:5]: @ilkkao 

if the db can't be reached for whatever reason, the error is shown.

wrong host:
```
mysqladmin: connect to server at 'db' failed
error: 'Can't connect to server on 'db' (115)'
Check that mariadbd is running and that the socket: '/run/mysqld/mysqld.sock' exists!
```

wrong credentials:
```
mysqladmin: connect to server at 'db' failed
error: 'Access denied for user 'mautic_db_users'@'172.22.0.3' (using password: YES)'
mysqladmin: connect to server at 'db' failed
error: 'Access denied for user 'mautic_db_users'@'172.22.0.3' (using password: YES)'
```

so I think that's fine?

File: common/entrypoint_mautic_web.sh
Comment (mollux) [reply common/entrypoint_mautic_web.sh:13]: @ilkkao 

as the same image is uses for multiple purposes (`web`, `cron`  and `worker`), the file can't be copied during image generation, as the ` /var/www/html/config/` folder is mounted as a volume.

This could lead to a started container overwriting the file in the volume, thus losing your local config.

Does that make sense?

File: common/entrypoint_mautic_web.sh
Comment (ilkkao) [reply common/entrypoint_mautic_web.sh:5]: Great, I wrongly assumed there's no output

File: common/entrypoint_mautic_web.sh
Comment (ilkkao) [reply common/entrypoint_mautic_web.sh:13]: Yes, I see. And users might modify this file (and add other files to the directory) so `/var/www/html/config/` needs to be a volume?

Comment (mollux) [reply common/entrypoint_mautic_web.sh:13]: indeed!

File: common/entrypoint_mautic_web.sh
Comment (mollux) [reply common/entrypoint_mautic_web.sh:5]: I also fixed the check, so it also takes into account incorrect host and port, see https://github.com/mautic/docker-mautic/pull/238/commits/8a58100de4eb83a09b87623cb8a1c0b7897bbb90

**Comment by vinyll** [comment]
You should update your user gitignore to avoid files and folders that are specific to your laptop:
```
echo ".DS_Store" >> ~/.gitignore
```

**Comment by vinyll** [comment]
I was trying to test out the fpm build with `docker build ./fpm` and got the error below.

```
 => CACHED [stage-1  5/13] RUN apt-get update && apt-get install --no-install-recommends -y     unzip libwebp-dev libzip-dev libfreetype6-dev libjpeg62-turbo-dev libpng-dev  0.0s
 => ERROR [stage-1  6/13] COPY ./common/php.ini /usr/local/etc/php/php.ini                                                                                                    0.0s
 => ERROR [stage-1  7/13] COPY ./common/docker-entrypoint.sh /entrypoint.sh                                                                                                   0.0s
 => ERROR [stage-1  8/13] COPY ./common/entrypoint_mautic_web.sh /entrypoint_mautic_web.sh                                                                                    0.0s
 => ERROR [stage-1  9/13] COPY ./common/entrypoint_mautic_cron.sh /entrypoint_mautic_cron.sh                                                                                  0.0s
 => ERROR [stage-1 10/13] COPY ./common/entrypoint_mautic_worker.sh /entrypoint_mautic_worker.sh                                                                              0.0s
 => CACHED [stage-1 11/13] RUN ["chmod", "+x", "/entrypoint.sh", "/entrypoint_mautic_web.sh", "/entrypoint_mautic_cron.sh", "/entrypoint_mautic_worker.sh"]                   0.0s
 => ERROR [stage-1 12/13] COPY ./common/supervisord.conf /etc/supervisor/conf.d/supervisord.conf                                                                              0.0s
------
 > [stage-1  6/13] COPY ./common/php.ini /usr/local/etc/php/php.ini:
------
------
 > [stage-1  7/13] COPY ./common/docker-entrypoint.sh /entrypoint.sh:
------
------
 > [stage-1  8/13] COPY ./common/entrypoint_mautic_web.sh /entrypoint_mautic_web.sh:
------
------
 > [stage-1  9/13] COPY ./common/entrypoint_mautic_cron.sh /entrypoint_mautic_cron.sh:
------
------
 > [stage-1 10/13] COPY ./common/entrypoint_mautic_worker.sh /entrypoint_mautic_worker.sh:
------
------
 > [stage-1 12/13] COPY ./common/supervisord.conf /etc/supervisor/conf.d/supervisord.conf:
------
Dockerfile:101
--------------------
  99 |         MAUTIC_WORKERS_CONSUME_FAILED=2
 100 |     
 101 | >>> COPY ./common/supervisord.conf /etc/supervisor/conf.d/supervisord.conf
 102 |     
 103 |     # Define Mautic volumes to persist data
--------------------
ERROR: failed to solve: failed to compute cache key: failed to calculate checksum of ref ce695d5e-5bdd-4b5d-9f15-e3f3d43c9763::pj788a3mlqfl1pan0fpsx0rx2: "/common/supervisord.conf": not found
```

despite the file exists:
```
ls -l common 
total 48
-rw-r--r--@ 1 vinyll  staff   297 Jan 11 20:35 docker-entrypoint.sh
-rw-r--r--@ 1 vinyll  staff   675 Jan 11 20:35 entrypoint_mautic_cron.sh
-rw-r--r--@ 1 vinyll  staff  1625 Jan 11 20:35 entrypoint_mautic_web.sh
-rw-r--r--@ 1 vinyll  staff    68 Jan 11 20:35 entrypoint_mautic_worker.sh
-rw-r--r--@ 1 vinyll  staff   316 Jan 11 20:35 php.ini
-rw-r--r--@ 1 vinyll  staff  1083 Jan 11 20:35 supervisord.conf
```

**Comment by mollux** [comment]
@vinyll did you check this PR or the main branch in the repo?
As https://github.com/mautic/docker-mautic/pull/238/files#diff-e2eb93a61ffd7877ea5c751abcb3a618e8e2e9a2073a27f66d4114fe10819f86 removes that file :)

**Comment by mollux** [comment]
@vinyll does `docker build -f fpm/Dockerfile .` work?

**Comment by vinyll** [comment]
> @vinyll did you check this PR or the main branch in the repo? As https://github.com/mautic/docker-mautic/pull/238/files#diff-e2eb93a61ffd7877ea5c751abcb3a618e8e2e9a2073a27f66d4114fe10819f86 removes that file :)

Right, I was just reviewing from the Github interface and that was just not totally clear to me.
It makes total sense then! :)

**Comment by vinyll** [comment]
> @vinyll does `docker build -f fpm/Dockerfile .` work?

Wow, it did!
The step `[mautic builder 4/7] RUN docker-php-ext-configure gd…` was time-taking (> 15 min) despite a high speed internet and laptop.

I had less success running it with the docker compose as the line
```
 => => # > mautic@0.0.0 build                                                                                                                                                      
 => => # > webpack --mode production  --config webpack.config.js                                                                                                                  
 => => # Modernizr build saved to node_modules/modernizr/modernizr-mautic-dist.js   
```

**Comment by ghostdevv** [comment]
Is there a planned release merge date for this PR? :pray:

**Comment by lumnn** [comment]
Either README needs changing to mention new `MYSQL_*` variables or they should have fall back to previous `MAUTIC_DB_*`

File: common/entrypoint_mautic_cron.sh
Comment (lumnn) [inline common/entrypoint_mautic_cron.sh:5]: /entrypoint_mautic_cron.sh: line 5: /opt/mautic/cron/mautic: No such file or directory

This is what I got when I'm running this entrypoint

Comment (lumnn) [reply common/entrypoint_mautic_cron.sh:5]: I was missing volume configuraion for this one to work. On the other hand it I believe it should work without volume attached as well.

File: common/entrypoint_mautic_cron.sh
Comment (lumnn) [inline common/entrypoint_mautic_cron.sh:10]: Is there a reason why only this is added to cron?

Documentation mentions at least about few other commands that are required
https://docs.mautic.org/en/5.x/configuration/cron_jobs.html

**Comment by odhekar** [comment]
It seems the REST API doesn't when I use the v5 image. Tried with both oauth2 and basic auth. 
I have tried `/contacts` and `/segments` and while both return correct total count, the rest of the response returns empty
```
{
    "total": "2",
    "contacts": {
        "1": {},
        "2": {}
    }
}
```

I am just trying out mautic first time and wanted to use latest images. am I missing a step to enable something for REST API in the container?

**Comment by odhekar** [comment]
went back to "old way" and 4.4.11 based image works fine for API calls.

**Comment by ghostdevv** [comment]
I think I also ran into this: https://github.com/mautic/mautic/issues/13243

**Comment by mollux** [comment]
@odhekar @ghostdevv the is indeed an issue with the API for composer based installs, and these Docker images use composer under the hood, so are also impacted.  I made [a PR to address this](https://github.com/mautic/mautic/pull/13272) that will be in the 5.0.3 release

File: common/entrypoint_mautic_cron.sh
Comment (mollux) [reply common/entrypoint_mautic_cron.sh:5]: @lumnn good point! I'll ensure the dir exists

Comment (mollux) [reply common/entrypoint_mautic_cron.sh:7]: @lumnn see https://github.com/mautic/docker-mautic/pull/238/commits/eb3ce519f423e256d534a7ac40c52a34df6cb2b3

**Comment by mollux** [comment]
I'm ready to merge this, so if anyone wan'ts to have a last look, now is the time.

The pipeline is running, and once successful, the images should pop up on https://github.com/mollux/docker-mautic/pkgs/container/docker-mautic and https://hub.docker.com/r/molluxmollux/test-mautic and 

Once merged, I will change the upstream registry to be https://hub.docker.com/r/mautic/mautic

File: common/entrypoint_mautic_cron.sh
Comment (mollux) [reply common/entrypoint_mautic_cron.sh:12]: @lumnn this was addressed, the default crons are present, and can be changed via a cron file

**Review by escopecz — APPROVED** [review]
This is amazing. Thank you Mattias! Can't wait to re-create my personal Mautic instances to this setup 👍

**Comment by mollux** [comment]
thanks! I'll merge, adapt the config to point to the correct upstreams and release the Docker images afterwards

**Pull request merged**
